influxdb3/batching/batcher.go

```diff
@@ -0,0 +1,153 @@
+
+The above copyright notice and this permission notice shall be included in
+all copies or substantial portions of the Software.
+
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+THE SOFTWARE.
+*/
+
+// Package batching provides a batcher to collect points and emit them as batches.
+package batching
+
+import (
+	"sync"
+
+	"github.com/InfluxCommunity/influxdb3-go/influxdb3"
+)
+
+// Option to adapt properties of a batcher
+type Option func(*Batcher)
+
+// WithSize changes the batch-size emitted by the batcher
+func WithSize(size int) Option {
+	return func(b *Batcher) {
+		b.size = size
+	}
+}
+
+// WithCapacity changes the initial capacity of the points buffer
+func WithCapacity(capacity int) Option {
+	return func(b *Batcher) {
+		b.capacity = capacity
+	}
+}
+
+// WithReadyCallback sets the function called when a new batch is ready. The
+// batcher will wait for the callback to finish, so please return as fast as
+// possible and move long-running processing to a  go-routine.
+func WithReadyCallback(f func()) Option {
+	return func(b *Batcher) {
+		b.callbackReady = f
+	}
+}
+
+// WithEmitCallback sets the function called when a new batch is ready with the
+// batch of points. The batcher will wait for the callback to finish, so please
+// return as fast as possible and move long-running processing to a go-routine.
+func WithEmitCallback(f func([]*influxdb3.Point)) Option {
+	return func(b *Batcher) {
+		b.callbackEmit = f
+	}
+}
+
+// DefaultBatchSize is the default number of points emitted
+const DefaultBatchSize = 1000
+
+// DefaultCapacity is the default initial capacity of the point buffer
+const DefaultCapacity = 2 * DefaultBatchSize
+
+// Batcher collects points and emits them as batches
+type Batcher struct {
+	size     int
+	capacity int
+
+	callbackReady func()
+	callbackEmit  func([]*influxdb3.Point)
+
+	points []*influxdb3.Point
+	sync.Mutex
+}
+
+// NewBatcher creates and initializes a new Batcher instance applying the
+// specified options. By default, a batch-size is DefaultBatchSize and the
+// initial capacity is DefaultCapacity.
+func NewBatcher(options ...Option) *Batcher {
+	// Set up a batcher with the default values
+	b := &Batcher{
+		size:     DefaultBatchSize,
+		capacity: DefaultCapacity,
+	}
+
+	// Apply the options
+	for _, o := range options {
+		o(b)
+	}
+
+	// Setup the internal data
+	b.points = make([]*influxdb3.Point, 0, b.capacity)
+
+	return b
+}
+
+// Add a metric to the batcher and call the given callbacks if any
+func (b *Batcher) Add(p *influxdb3.Point) {
```

Comment:
I was wondering if this could be a variadic function, so that `[]*influxdb3.Point` could be used as an argument as well, or whether we might want o support adding `[]*influxdb3.Point` in another function.

Reply:
Good point. Variadic parameters look good.